lib/cache/inventory_cache.go

```diff
@@ -0,0 +1,655 @@
+// 500          | 283              | 1.77s
+// 1,000        | 298              | 3.36s
+// 2,000        | 322              | 6.21s
+// 4,000        | 363              | 11.02s
+// 8,000        | 433              | 18.47s
+// 32,000       | 789              | 40.56s
+// 64,000       | 1219             | 52.5s
+// 128,000      | 2035             | 1m03s
+// 256,000      | 3605             | 1m11s
+func calculateReadsPerSecond(clusterSize int) int {
+	// minimumComponent is the minimum value of reads per second we never want to drop below.
+	const minimumComponent = 256
+
+	// linearComponent ensures we stay under a worst-case upper bound init time of 90s.
+	linearComponent := clusterSize / 90
+
+	// subLinearComponent ensures that growth is sub-linear across most reasonable cluster sizes.
+	subLinearComponent := int(math.Sqrt(float64(clusterSize)))
+
+	return minimumComponent + linearComponent + subLinearComponent
+}
+
+// initializeAndWatchWithRetry runs initializeAndWatch with a retry every 10 seconds if it fails.
+func (ic *InventoryCache) initializeAndWatchWithRetry(ctx context.Context) {
+	defer close(ic.initDone)
+
+	const retryInterval = 10 * time.Second
+
+	for {
+		ic.cfg.Logger.DebugContext(ctx, "Attempting to initialize inventory cache")
+
+		// Attempt to initialize and watch
+		if err := ic.initializeAndWatch(ctx); err != nil {
+			if ctx.Err() != nil {
+				return
+			}
+
+			ic.cfg.Logger.WarnContext(ctx, "Failed to initialize inventory cache, retrying in 10 seconds",
+				"error", err)
+
+			// Wait before retrying
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(retryInterval):
+			}
+			continue
+		}
+
+		return
+	}
+}
+
+// initializeAndWatch initializes the inventory cache and begins watching for instance and bot_instance backend events.
+func (ic *InventoryCache) initializeAndWatch(ctx context.Context) error {
+	// Wait for primary cache to be ready.
+	if err := ic.waitForPrimaryCacheInit(ctx); err != nil {
+		return trace.Wrap(err, "Failed to wait for primary cache init")
+	}
+
+	// Setup the backend watcher.
+	watcher, err := ic.setupWatcher(ctx)
+	if err != nil {
+		return trace.Wrap(err, "Failed to set up backend watcher")
+	}
+	defer watcher.Close()
+
+	// Wait for the watcher to be ready.
+	if err := ic.waitForWatcherInit(ctx, watcher); err != nil {
+		return trace.Wrap(err, "Failed to wait for watcher init")
+	}
+
+	// Calculate the rate limit to use.
+	primaryCacheSize := ic.cfg.PrimaryCache.GetCacheSize()
+	readsPerSecond := calculateReadsPerSecond(primaryCacheSize)
+
+	// Populate the cache with teleport instance and bot instances.
+	if err := ic.populateCache(ctx, readsPerSecond); err != nil {
+		return trace.Wrap(err, "failed to populate inventory cache")
+	}
+
+	// Mark cache as healthy.
+	ic.healthy.Store(true)
+
+	// This runs infinitely until the context is cancelled, so the return below won't be hit until shutdown.
+	ic.processEvents(ctx, watcher)
```

Comment:
If we have enough items that we need to rate limit our fetch, don't we risk having so many items that the watcher event buffer fills up if we don't drain it?